### DISCUSSION TITLE
How to run included .qmds interactively?

### DISCUSSION THREAD

**Discussion by edalfon** [quarto-dev/quarto-cli#7872] [Q&A]
### Description

Hi, we have a [Quarto project](https://quarto.org/docs/projects/quarto-projects.html) where we use [includes](https://quarto.org/docs/authoring/includes.html) to make sure all chapters run some shared code. So basically at the beginning of several files we have something like:
```qmd
{{< include _startup_code.qmd >}}
```
This works just fine when you render it with `quarto render myproject`, but to try things out interactively, it's a bit inconvenient. We need to run the code `_startup_code.qmd` before keep working on the chapter, and there does not seem to be a shortcut to do it (in vscode with the quarto extension installed and running the code in an ipython kernel), so basically we would need to manually open the file and run the code and then go back to the chapter to try things out.

Any ideas how to do this?, any easy way to instruct vscode to do it, and sort-of, shift-enter our way to automagically run all the code in the included file? Any ideas are greatly appreciated.

Note: perhaps we are abusing the [includes](https://quarto.org/docs/authoring/includes.html) functionality to avoid repeating ourselves too much and keep this relatively small project maintainable. So, sorry for that. Just to say that for larger/more complex projects we would use other orchestration tools (e.g. targets in R, or prefect, dvc, ploomber for python projects) and probably would not rely on includes this way.

**Comment by cscheid** [comment]
It's a valid request. One possibility would be for VS Code and RStudio to detect include shortcodes, and interpret the "execute code cell" to mean "expand all includes and execute all resulting cells". It's a bit of work for us to set it up, but I think it's the right thing to do.

**Reply by edalfon** [reply]
thx, that sounds great!

**Reply by cderv** [reply]
Just for reference Idea tracked at 
* https://github.com/quarto-dev/quarto/issues/411

**Comment by gabrielparriaux** [comment]
Hi there,

I have the same kind of request, I think. 

I built a Quarto document with around 5000 lines of code (mostly R, also a few markdown) to analyse the transcript of teachers’ discourse in class. As I have 27 classes, I have to reuse this same Quarto document in 27 different projects so that I have one project for each class (with its own import and export, and so on). 

I have first worked copying and pasting code from one project to the other, but with the number of classes increasing, it’s becoming insane when I want to update my code. I have to do it in all projects! And I can’t reuse the exact same code for the whole file completely, because there are some specific adaptations that I have to do for each class.

So I discovered the opportunity to use `{{< include _my_include.qmd >}}` and thought it was a perfect solution for me. But after a try, I’m ending with a main .qmd file and six different include files that come at different places in the main file. 

It works well, but the problem is when I want to execute the whole code… I have to execute chunks 1-3 in the main file, then the first include, then chunks 4-5 in the main file, then the second include, and so on… It’s not practical at all and I’m questioning if it’s a good idea to use includes for my case. Maybe it’s better to stick with my original copy paste of the whole file? Or is there another way around that I didn’t see?

If we could automate the execution of the whole code, with the includes, and in the good order with just one click, it would be great!

I’m working in RStudio.

Thank you for helping!

**Reply by mcanouil** [reply]
The previous answer is still valid.

One alternative in your case, could be to write a format extension where the template is your big set of code (https://quarto.org/docs/extensions/listing-formats.html).
Then in a new project you would:
```sh
quarto use template user/repo
```

**Reply by gabrielparriaux** [reply]
Thank you @mcanouil for the proposition.

I’m not sure that it would fit for me, because in my case I’m very frequently updating the code for all my projects (adding new analysis and so on). As far as I understand the way templates are working, it helps to start with the code instead of nothing, but then it doesn’t propose any help in the case you want to update the code for all your projects in one step, right?

Anyway, it was very interesting to discover Quarto extensions as I didn’t know them and I already decided to use some of them!

**Reply by mcanouil** [reply]
There is nothing that will update something everywhere.
So, you need to figure out your structure.
You could have one single document and use parameters to targets different directories/projects.
You could also use hardlinks of one file into every projects.

**Reply by gabrielparriaux** [reply]
The solution with `{{< include _my_include.qmd >}}` is fine for me, I think. I can centralise pieces of code in a folder shared by all my projects. 

I can render with one click and everything’s fine. It’s only when I was to run code chunks that it’s a little tricky to be done manually… so, if some day this possibility of running all the code cells also from the included files exists, then I’ll be perfectly happy!
